include/forcing/ForcingsEngine.hpp

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <vector>
+#include <boost/core/span.hpp>
+
+#include "Bmi_Py_Adapter.hpp"
+#include <utilities/mdarray.hpp>
+
+namespace data_access {
+
+struct ForcingsEngine
+{
+    using size_type                 = std::size_t;
+    using clock_type                = std::chrono::system_clock;
+    static constexpr auto bad_index = static_cast<size_type>(-1);
+
+    ForcingsEngine(const std::string& init, size_type time_start, size_type time_end);
+
+    ~ForcingsEngine();
+
+    /**
+     * @brief Get the output variables associated with this instance of the
+     *        Forcings Engine
+     * 
+     * @return boost::span<const std::string> 
+     */
+    boost::span<const std::string> outputs() const noexcept;
+
+    /**
+     * @brief Get the beginning time point for this instance of the
+     *        Forcings Engine
+     * 
+     * @return const clock_type::time_point& 
+     */
+    const clock_type::time_point& time_begin() const noexcept;
+
+    /**
+     * @brief Get the ending time point for this instance of the
+     *        Forcings Engine
+     * 
+     * @return const clock_type::time_point& 
+     */
+    const clock_type::time_point& time_end() const noexcept;
+
+    /**
+     * @brief Get the stride/time step this instance of the
+     *        Forcings Engine follows.
+     * 
+     * @return const clock_type::duration& 
+     */
+    const clock_type::duration& time_step() const noexcept;
+
+    /**
+     * @brief Get the index for a C-style epoch within this instance of the
+     *        Forcings Engine's temporal domain.
+     * 
+     * @param ctime C-style epoch
+     * @return size_type, where the index represents a point between time_begin() and time_end() by time_step().
+     */
+    size_type time_index(time_t ctime) const noexcept;
+
+    /**
+     * @brief Get the index in `CAT-ID` for a given divide in the instance cache.
+     * @note The `CAT-ID` output variable uses integer values instead of strings.
+     * 
+     * @param divide_id A hydrofabric divide ID, i.e. "cat-*"
+     * @return size_type 
+     */
+    size_type divide_index(const std::string& divide_id) const noexcept;
+
+    /**
+     * @brief Get the index of a variable in the instance cache.
+     * 
+     * @param variable 
+     * @return size_type 
+     */
+    size_type variable_index(const std::string& variable) const noexcept;
```

Comment:
There's public functions for indices of time, divide, and variable, and then the actual data access function `at()` takes the key values, and not the indices. What's the deal with that?

Reply:
`at` uses these three functions, I intended to move them to the private interface but neglected that.

Reply:
Honestly, I'd kinda rather `at` took (or had an overload that took) the index values instead, and callers kept some awareness of indices. Juggling strings and system time types is kinda silly for a core data structure.